interfaces/sourcegen/sourcegen/csharp/_SourceGenerator.py

```diff
@@ -0,0 +1,309 @@
+# This file is part of Cantera. See License.txt in the top-level directory or
+# at https://cantera.org/license.txt for license and copyright information.
+
+from dataclasses import dataclass
+import os
+import re
+import typing
+
+from .._helpers import normalize_indent, get_preamble
+from .._types import *
+from .._types import _unpack
+
+
+@dataclass
+class _CsFunc:
+    """ Represents a C# method """
+
+    ret_type: str
+    name: str
+    params: list[Param]
+    del_clazz: typing.Optional[str]
+
+
+    def __iter__(self):
+        return _unpack(self)
+
+
+class SourceGenerator(SourceGeneratorBase):
+    _prolog = normalize_indent("""
+        [DllImport(LibFile)]
+        public static extern
+    """)
+
+    _type_map = {
+        'const char*': 'string',
+        'const double*': 'double[]',
+        'size_t': 'nuint',
+        'char*': 'byte*'
+    }
+
+    _prop_type_map = {
+        'byte*': 'string',
+        'double*': 'double[]'
+    }
+
+    _preamble = '/*\n' + get_preamble() + '*/'
+
+    @staticmethod
+    def _join_params(params):
+        return ', '.join((p.p_type + ' ' + p.name for p in params))
+
+
+    @classmethod
+    def _get_function_text(cls, function):
+        ref_type, name, params, _ = function
+        is_unsafe = any((p.p_type.endswith('*') for p in params))
+        params_text = cls._join_params(params)
+        if is_unsafe:
+            return f'{cls._prolog} unsafe {ref_type} {name}({params_text});'
+        else:
+            return f'{cls._prolog} {ref_type} {name}({params_text});'
+
+
+    @staticmethod
+    def _get_base_handle_text(handle):
+        name, del_clazz = handle
+
+        handle = normalize_indent(f'''
+            class {del_clazz} : CanteraHandle
+            {{
+                protected override bool ReleaseHandle() =>
+                    LibCantera.{name}(Value) == InteropConsts.Success;
+            }}
+        ''')
+
+        return handle
+
+
+    @staticmethod
+    def _get_derived_handle_text(derived):
+        derived, base = derived
+
+        derived = f'''class {derived} : {base} {{ }}'''
+
+        return derived
+
+
+    def __init__(self, out_dir: str, config: dict):
+        self._out_dir = out_dir
+        self._config = config
+        # as different clib files are passed in, the known funcs are added to this dict
+        self._known_funcs = {}
+
+        os.makedirs(out_dir, exist_ok=True)
+
+
+    def _convert_func(self, parsed: Func):
+        ret_type, name, params = parsed
+        clazz, method = name.split('_', 1)
+
+        #copy the params list
+        params = list(params)
```

Comment:
Why do you need to copy the `params`? This won't create copies of the underlying `Param` instances, to do that you need `copy.deepcopy`.

Reply:
This method maps C parameters to C# parameters for the so-called P/Invoke signatures on the C# side. In many cases, they are identical, but this method looks for the cases where I need to or can change the C# side for syntax, safety, or performance reasons. For example, C `const char *` (i.e. sending string data to libcantera) can become .NET's `String` type*. 

The incoming params list onstensibly belongs to the C function, so I want a new list for the C# function. I've added `frozen=True` to all the `dataclass`es, so I can be sure I'm not accidentally modifying an individual item, just replacing it completely if needed.

-----

* As part of this process, I'm hoping to document more about how I devised the interop layer, and the magic of making it high-performance when needed vs. simple when possible.

Reply:
> The incoming params list onstensibly belongs to the C function, so I want a new list for the C# function

This still doesn't answer the _why_ for me, though 😄 To be clear, my concern is for the long term maintainability of this code... if I come back to it in 6 months or 2 years, I won't know why this list is being copied and I might change the behavior by accident and introduce a bug.

That said, this is in the csharp-specific part of the code, so maybe it's alright with just a descriptive comment about why this is happening.

Reply:
OK so the why is that the C# signature will need params just like the C signature, and further, some of them will be syntatically identical, so I can reuse much of the same list, but I also want to replace items that need to special handling. But the orignal list is still referenced by the C func, so I don’t want to mutate it.

So it sounds like the thing here is to rename the local variable and/or add more info to the comments.

Reply:
Added a comment to the source to clarify.